tower-http/src/sensitive_header.rs

```diff
@@ -0,0 +1,224 @@
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        if let Some(value) = req.headers_mut().get_mut(&self.header) {
+            value.set_sensitive(true);
+        }
+
+        self.inner.call(req)
+    }
+}
+
+/// Mark a response header as [sensitive].
+///
+/// [sensitive]: https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.set_sensitive
+#[derive(Clone, Debug)]
+pub struct SetSensitiveResponseHeaderLayer {
+    header: HeaderName,
+}
+
+impl SetSensitiveResponseHeaderLayer {
+    /// Create a new [`SetSensitiveResponseHeaderLayer`].
+    pub fn new(header: HeaderName) -> Self {
+        Self { header }
+    }
+}
+
+impl<S> Layer<S> for SetSensitiveResponseHeaderLayer {
+    type Service = SetSensitiveResponseHeader<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetSensitiveResponseHeader {
+            inner,
+            header: self.header.clone(),
+        }
+    }
+}
+
+/// Mark a response header as [sensitive].
+///
+/// [sensitive]: https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.set_sensitive
+#[derive(Clone, Debug)]
+pub struct SetSensitiveResponseHeader<S> {
+    inner: S,
+    header: HeaderName,
+}
+
+impl<S> SetSensitiveResponseHeader<S> {
+    /// Create a new [`SetSensitiveResponseHeader`].
+    pub fn new(inner: S, header: HeaderName) -> Self {
+        Self { inner, header }
+    }
+
+    /// Gets a reference to the underlying service.
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying service.
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consumes `self`, returning the underlying service.
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+
+impl<ReqBody, ResBody, S> Service<Request<ReqBody>> for SetSensitiveResponseHeader<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = SetSensitiveResponseHeaderResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        SetSensitiveResponseHeaderResponseFuture {
+            future: self.inner.call(req),
+            header: Some(self.header.clone()),
+        }
+    }
+}
+
+/// Response future for [`SetSensitiveResponseHeader`].
+#[pin_project]
+#[derive(Debug)]
+pub struct SetSensitiveResponseHeaderResponseFuture<F> {
+    #[pin]
+    future: F,
+    header: Option<HeaderName>,
```

Comment:
i'm not sure if i get why this needs to be an option. we only need to borrow it to call `get_mut`, we don't need to move it by value --- can't this just be a `HeaderName`? if we did that, we could avoid the `take()` and `unwrap()` calls in the `Future` impl, which would make the code a little shorter (and not generate a panic that never actually occurs)...

Reply:
Good catch. `res.headers_mut().get_mut(&*this.header)` did the trick.